File: inventory-oms-poc/algorithm-review/python/oms_algorithms.py
"""Algorithm patterns translated into OMS, payment and data-governance examples.

The functions are deliberately pure or local-memory examples. They explain the
algorithmic invariant; production adapters must add persistence, concurrency
control, authentication, audit and failure recovery.
"""

from __future__ import annotations

from bisect import bisect_right
from collections import OrderedDict, deque
from dataclasses import dataclass
import heapq
from typing import Deque, Iterable, Mapping, Sequence


@dataclass(frozen=True)
class RequestEvent:
    """One already time-ordered request observation."""

    key: str
    timestamp_ms: int


def deduplicate_keys(keys: Iterable[str]) -> list[bool]:
    """Return whether each key is the first occurrence.

    This is the HashSet part of an idempotency check.

    Time: O(n) average. Space: O(u), where u is the number of unique keys.
    Production: use a database unique constraint or an atomic Redis operation
    across replicas; a process-local set is only an optimization.
    """

    seen: set[str] = set()
    first_seen: list[bool] = []
    for key in keys:
        first_seen.append(key not in seen)
        seen.add(key)
    return first_seen


def sliding_window_allow(
    events: Sequence[RequestEvent],
    limit: int,
    window_ms: int,
) -> list[bool]:
    """Decide whether each request is allowed in a fixed-size time window.

    The input must be non-decreasing by timestamp for each key. The queue keeps
    only timestamps that can still affect the current decision.

    Time: O(n) amortized because each timestamp enters and leaves once.
    Space: O(n) worst case for timestamps inside active windows.
    Production: implement the same invariant atomically with Redis ZSET + Lua
    or a gateway-native rate limiter.
    """

    if limit <= 0 or window_ms <= 0:
        raise ValueError("limit and window_ms must be positive")

    windows: dict[str, Deque[int]] = {}
    decisions: list[bool] = []

    for event in events:
        queue = windows.setdefault(event.key, deque())
        cutoff = event.timestamp_ms - window_ms
        while queue and queue[0] <= cutoff:
            queue.popleft()

        allowed = len(queue) < limit
        decisions.append(allowed)
        if allowed:
            queue.append(event.timestamp_ms)

    return decisions


@dataclass(frozen=True, order=True)
class SettlementRecord:
    """Minimal provider settlement record used by reconciliation."""

    occurred_at_ms: int
    provider: str
    transaction_id: str
    amount_cents: int
    status: str


def merge_settlement_streams(
    streams: Sequence[Sequence[SettlementRecord]],
) -> list[SettlementRecord]:
    """Merge K individually sorted settlement streams.

    The heap contains at most one head record from each stream. The tie-breaker
    fields make output deterministic, which is important for replay and audit.

    Time: O(n log k), where n is total records and k is stream count.
    Space: O(k) heap space, excluding the returned output.
    """

    heap: list[tuple[SettlementRecord, int, int]] = []
    for stream_index, stream in enumerate(streams):
        if stream:
            heapq.heappush(heap, (stream[0], stream_index, 0))

    merged: list[SettlementRecord] = []
    while heap:
        record, stream_index, record_index = heapq.heappop(heap)
        merged.append(record)

        next_index = record_index + 1
        stream = streams[stream_index]
        if next_index < len(stream):
            heapq.heappush(heap, (stream[next_index], stream_index, next_index))

    return merged


def topological_order(
    nodes: Iterable[str],
    dependencies: Iterable[tuple[str, str]],
) -> list[str]:
    """Return a valid order for edges (prerequisite, dependent).

    Kahn's BFS algorithm repeatedly removes zero-indegree nodes. If not all
    nodes are emitted, the graph contains a cycle.

    Time: O(V + E). Space: O(V + E).
    Production: use this for dependency linting and explainability; Terraform,
    Argo CD and Kubernetes remain the authoritative execution controllers.
    """

    node_set = set(nodes)
    adjacency: dict[str, list[str]] = {node: [] for node in node_set}
    indegree = {node: 0 for node in node_set}

    for prerequisite, dependent in dependencies:
        if prerequisite not in node_set or dependent not in node_set:
            raise ValueError("dependency references an unknown node")
        adjacency[prerequisite].append(dependent)
        indegree[dependent] += 1

    ready = deque(sorted(node for node in node_set if indegree[node] == 0))
    result: list[str] = []

    while ready:
        node = ready.popleft()
        result.append(node)
        for dependent in sorted(adjacency[node]):
            indegree[dependent] -= 1
            if indegree[dependent] == 0:
                ready.append(dependent)

    if len(result) != len(node_set):
        raise ValueError("dependency graph contains a cycle")
    return result


@dataclass(frozen=True)
class DataFlowNode:
    """A node in a governed data-flow graph."""

    name: str
    region: str
    data_class: str


def find_cross_region_violation(
    start: str,
    graph: Mapping[str, Sequence[str]],
    nodes: Mapping[str, DataFlowNode],
    allowed_regions: set[str],
) -> list[str] | None:
    """Find one shortest path that leaves the allowed data region.

    This is a BFS because the first violation path is useful in an audit report.
    It detects a path; it does not grant access, approve cross-border transfer,
    or replace policy-as-code.

    Time: O(V + E). Space: O(V).
    """

    if start not in nodes:
        raise ValueError(f"unknown start node: {start}")

    queue: deque[str] = deque([start])
    parent: dict[str, str | None] = {start: None}

    while queue:
        current = queue.popleft()
        for neighbor in graph.get(current, ()):
            if neighbor not in nodes:
                raise ValueError(f"unknown graph node: {neighbor}")
            if neighbor in parent:
                continue

            parent[neighbor] = current
            if nodes[neighbor].region not in allowed_regions:
                path: list[str] = []
                cursor: str | None = neighbor
                while cursor is not None:
                    path.append(cursor)
                    cursor = parent[cursor]
                return list(reversed(path))
            queue.append(neighbor)

    return None


class LruCache:
    """O(1) average get/put cache using a dict plus an order list.

    OrderedDict is used here only to keep the sample compact. The interview
    invariant is still HashMap + doubly linked list. Production stores only a
    non-authoritative local copy of configuration or risk rules.
    """

    def __init__(self, capacity: int) -> None:
        if capacity <= 0:
            raise ValueError("capacity must be positive")
        self._capacity = capacity
        self._items: OrderedDict[str, str] = OrderedDict()

    def get(self, key: str) -> str | None:
        if key not in self._items:
            return None
        self._items.move_to_end(key)
        return self._items[key]

    def put(self, key: str, value: str) -> None:
        self._items[key] = value
        self._items.move_to_end(key)
        if len(self._items) > self._capacity:
            self._items.popitem(last=False)

    def __len__(self) -> int:
        return len(self._items)


@dataclass(frozen=True)
class ChannelOption:
    units: int
    fee_cents: int


def min_channel_cost(target_units: int, options: Sequence[ChannelOption]) -> int | None:
    """Find the minimum fee for an exact discrete amount.

    This is an unbounded DP example for planning package/channel choices.
    It is not a provider authorization algorithm and does not bypass provider
    limits, currency rules, settlement risk or compliance controls.

    Time: O(target_units * number_of_options).
    Space: O(target_units).
    """

    if target_units < 0 or any(
        option.units <= 0 or option.fee_cents < 0 for option in options
    ):
        raise ValueError("target and options must be non-negative and usable")

    infinity = 10**18
    dp = [0] + [infinity] * target_units
    for current in range(1, target_units + 1):
        for option in options:
            if option.units <= current:
                dp[current] = min(
                    dp[current],
                    dp[current - option.units] + option.fee_cents,
                )

    return None if dp[target_units] == infinity else dp[target_units]


def merge_intervals(intervals: Sequence[tuple[int, int]]) -> list[tuple[int, int]]:
    """Merge overlapping deployment freeze or maintenance windows.

    Time: O(n log n) for sorting. Space: O(n) for the result.
    """

    if not intervals:
        return []

    ordered = sorted(intervals)
    merged: list[list[int]] = [[ordered[0][0], ordered[0][1]]]
    for start, end in ordered[1:]:
        if start <= merged[-1][1]:
            merged[-1][1] = max(merged[-1][1], end)
        else:
            merged.append([start, end])
    return [(start, end) for start, end in merged]


@dataclass(frozen=True)
class RankingEntry:
    """One already sorted member from a leaderboard shard."""

    score: int
    member: str


def merge_sharded_top_k(
    shards: Sequence[Sequence[RankingEntry]],
    limit: int,
) -> list[RankingEntry]:
    """Merge per-shard descending leaderboards and return the global Top-K.

    Each shard must be sorted by ``score`` descending and ``member`` ascending.
    Only one head per shard is kept in the heap, so the complexity is
    O(K log S), where S is the number of shards. This models the live-ranking
    design from the recording: query each Redis ZSET shard for its local Top-K,
    then merge at the application layer.
    """

    if limit <= 0:
        raise ValueError("limit must be positive")

    heap: list[tuple[int, str, int, int]] = []
    for shard_index, shard in enumerate(shards):
        if shard:
            first = shard[0]
            heapq.heappush(heap, (-first.score, first.member, shard_index, 0))

    result: list[RankingEntry] = []
    while heap and len(result) < limit:
        _, _, shard_index, entry_index = heapq.heappop(heap)
        entry = shards[shard_index][entry_index]
        result.append(entry)
        next_index = entry_index + 1
        if next_index < len(shards[shard_index]):
            next_entry = shards[shard_index][next_index]
            heapq.heappush(
                heap,
                (-next_entry.score, next_entry.member, shard_index, next_index),
            )
    return result


@dataclass(frozen=True, order=True)
class CursorRecord:
    """A stable ascending keyset-pagination record."""

    sort_key: int
    record_id: str


def keyset_page(
    records: Sequence[CursorRecord],
    after: CursorRecord | None,
    limit: int,
) -> tuple[list[CursorRecord], CursorRecord | None]:
    """Return one page after a stable ``(sort_key, record_id)`` cursor.

    ``records`` must already be ordered by the cursor fields, as a database
    query with ``ORDER BY sort_key, record_id`` would be. A real repository
    should apply ``WHERE (sort_key, record_id) > (:sort_key, :record_id)``
    against the matching composite index rather than scan and discard rows.
    """

    if limit <= 0:
        raise ValueError("limit must be positive")

    start = 0 if after is None else bisect_right(records, after)
    page = list(records[start : start + limit])
    next_cursor = page[-1] if start + len(page) < len(records) and page else None
    return page, next_cursor


def _demo() -> None:
    assert deduplicate_keys(["a", "a", "b"]) == [True, False, True]
    assert sliding_window_allow(
        [
            RequestEvent("merchant-1", 0),
            RequestEvent("merchant-1", 100),
            RequestEvent("merchant-1", 200),
        ],
        limit=2,
        window_ms=1_000,
    ) == [True, True, False]

    streams = [
        [SettlementRecord(1, "wx", "w-1", 100, "CAPTURED")],
        [SettlementRecord(2, "ali", "a-1", 100, "CAPTURED")],
    ]
    assert [record.transaction_id for record in merge_settlement_streams(streams)] == [
        "w-1",
        "a-1",
    ]

    assert topological_order(
        ["vpc", "eks", "aurora"], [("vpc", "eks"), ("eks", "aurora")]
    ) == ["vpc", "eks", "aurora"]

    graph = {
        "cn-db": ["cn-kafka"],
        "cn-kafka": ["sg-databricks"],
        "sg-databricks": [],
    }
    nodes = {
        "cn-db": DataFlowNode("cn-db", "cn", "raw_personal"),
        "cn-kafka": DataFlowNode("cn-kafka", "cn", "raw_personal"),
        "sg-databricks": DataFlowNode("sg-databricks", "sg", "derived"),
    }
    assert find_cross_region_violation(
        "cn-db", graph, nodes, {"cn"}
    ) == ["cn-db", "cn-kafka", "sg-databricks"]

    cache = LruCache(1)
    cache.put("risk-rule", "v1")
    assert cache.get("risk-rule") == "v1"
    cache.put("other", "v2")
    assert cache.get("risk-rule") is None

    assert min_channel_cost(
        6, [ChannelOption(units=3, fee_cents=5), ChannelOption(units=2, fee_cents=4)]
    ) == 10
    assert merge_intervals([(1, 3), (2, 5), (8, 9)]) == [(1, 5), (8, 9)]

    leaderboard = merge_sharded_top_k(
        [
            [RankingEntry(100, "u-1"), RankingEntry(80, "u-4")],
            [RankingEntry(100, "u-2"), RankingEntry(90, "u-3")],
        ],
        3,
    )
    assert [(entry.score, entry.member) for entry in leaderboard] == [
        (100, "u-1"),
        (100, "u-2"),
        (90, "u-3"),
    ]

    records = [CursorRecord(10, "a"), CursorRecord(10, "b"), CursorRecord(11, "c")]
    page, cursor = keyset_page(records, CursorRecord(10, "a"), 1)
    assert page == [CursorRecord(10, "b")] and cursor == CursorRecord(10, "b")


if __name__ == "__main__":
    _demo()
    print("Python algorithm review demo passed")

File: inventory-oms-poc/algorithm-review/go/oms_algorithms.go
package algorithms

import (
	"container/heap"
	"container/list"
	"errors"
	"sort"
)

// RequestEvent is a time-ordered request observation.
type RequestEvent struct {
	Key         string
	TimestampMs int64
}

// DeduplicateKeys returns true only for the first occurrence of each key.
//
// Time: O(n) average. Space: O(u), where u is the number of unique keys.
func DeduplicateKeys(keys []string) []bool {
	seen := make(map[string]struct{}, len(keys))
	result := make([]bool, 0, len(keys))
	for _, key := range keys {
		_, exists := seen[key]
		result = append(result, !exists)
		seen[key] = struct{}{}
	}
	return result
}

// SlidingWindowAllow applies a per-key fixed-size window.
//
// Events for the same key must be non-decreasing by TimestampMs. Each
// timestamp is appended and removed once, so the amortized time is O(n).
// Production should move the atomic state to Redis or the API gateway.
func SlidingWindowAllow(events []RequestEvent, limit int, windowMs int64) ([]bool, error) {
	if limit <= 0 || windowMs <= 0 {
		return nil, errors.New("limit and windowMs must be positive")
	}

	timestamps := make(map[string][]int64)
	heads := make(map[string]int)
	result := make([]bool, 0, len(events))

	for _, event := range events {
		queue := timestamps[event.Key]
		head := heads[event.Key]
		cutoff := event.TimestampMs - windowMs
		for head < len(queue) && queue[head] <= cutoff {
			head++
		}

		allowed := len(queue)-head < limit
		result = append(result, allowed)
		if allowed {
			queue = append(queue, event.TimestampMs)
		}
		timestamps[event.Key] = queue
		heads[event.Key] = head
	}
	return result, nil
}

// SettlementRecord is the minimum provider record needed for reconciliation.
type SettlementRecord struct {
	OccurredAtMs int64
	Provider     string
	Transaction  string
	AmountCents  int64
	Status       string
}

type mergeItem struct {
	record      SettlementRecord
	streamIndex int
	recordIndex int
}

type settlementHeap []mergeItem

func (h settlementHeap) Len() int { return len(h) }
func (h settlementHeap) Less(i, j int) bool {
	left, right := h[i].record, h[j].record
	if left.OccurredAtMs != right.OccurredAtMs {
		return left.OccurredAtMs < right.OccurredAtMs
	}
	if left.Provider != right.Provider {
		return left.Provider < right.Provider
	}
	return left.Transaction < right.Transaction
}
func (h settlementHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
func (h *settlementHeap) Push(value interface{}) {
	*h = append(*h, value.(mergeItem))
}
func (h *settlementHeap) Pop() interface{} {
	old := *h
	last := len(old) - 1
	value := old[last]
	*h = old[:last]
	return value
}

// MergeSettlementStreams merges K individually sorted provider streams.
//
// Time: O(n log k); heap space: O(k).
func MergeSettlementStreams(streams [][]SettlementRecord) []SettlementRecord {
	queue := &settlementHeap{}
	heap.Init(queue)
	for streamIndex, stream := range streams {
		if len(stream) > 0 {
			heap.Push(queue, mergeItem{stream[0], streamIndex, 0})
		}
	}

	result := make([]SettlementRecord, 0)
	for queue.Len() > 0 {
		item := heap.Pop(queue).(mergeItem)
		result = append(result, item.record)

		nextIndex := item.recordIndex + 1
		if nextIndex < len(streams[item.streamIndex]) {
			heap.Push(queue, mergeItem{
				streams[item.streamIndex][nextIndex],
				item.streamIndex,
				nextIndex,
			})
		}
	}
	return result
}

// RankingEntry is one member from a descending leaderboard shard.
type RankingEntry struct {
	Score  int64
	Member string
}

type rankingItem struct {
	entry       RankingEntry
	shardIndex  int
	entryIndex  int
}

type rankingHeap []rankingItem

func (h rankingHeap) Len() int { return len(h) }
func (h rankingHeap) Less(i, j int) bool {
	if h[i].entry.Score != h[j].entry.Score {
		return h[i].entry.Score > h[j].entry.Score
	}
	if h[i].entry.Member != h[j].entry.Member {
		return h[i].entry.Member < h[j].entry.Member
	}
	return h[i].shardIndex < h[j].shardIndex
}
func (h rankingHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
func (h *rankingHeap) Push(value interface{}) { *h = append(*h, value.(rankingItem)) }
func (h *rankingHeap) Pop() interface{} {
	old := *h
	last := len(old) - 1
	value := old[last]
	*h = old[:last]
	return value
}

// MergeShardedTopK merges already sorted per-shard leaderboards.
// Time: O(K log S), where S is the number of shards.
func MergeShardedTopK(shards [][]RankingEntry, limit int) []RankingEntry {
	if limit <= 0 {
		return nil
	}
	queue := &rankingHeap{}
	heap.Init(queue)
	for shardIndex, shard := range shards {
		if len(shard) > 0 {
			heap.Push(queue, rankingItem{shard[0], shardIndex, 0})
		}
	}

	result := make([]RankingEntry, 0, limit)
	for queue.Len() > 0 && len(result) < limit {
		item := heap.Pop(queue).(rankingItem)
		result = append(result, item.entry)
		nextIndex := item.entryIndex + 1
		if nextIndex < len(shards[item.shardIndex]) {
			heap.Push(queue, rankingItem{
				shards[item.shardIndex][nextIndex], item.shardIndex, nextIndex,
			})
		}
	}
	return result
}

// CursorRecord is a stable ascending keyset-pagination key.
type CursorRecord struct {
	SortKey  int64
	RecordID string
}

// KeysetPage returns records after a cursor and the next cursor, if any.
// The input must be sorted by SortKey then RecordID. A database adapter should
// apply the same tuple predicate with a matching composite index.
func KeysetPage(records []CursorRecord, after *CursorRecord, limit int) ([]CursorRecord, *CursorRecord, error) {
	if limit <= 0 {
		return nil, nil, errors.New("limit must be positive")
	}
	start := 0
	if after != nil {
		start = sort.Search(len(records), func(index int) bool {
			current := records[index]
			return current.SortKey > after.SortKey ||
				(current.SortKey == after.SortKey && current.RecordID > after.RecordID)
		})
	}
	end := start + limit
	if end > len(records) {
		end = len(records)
	}
	page := append([]CursorRecord(nil), records[start:end]...)
	if end == len(records) || len(page) == 0 {
		return page, nil, nil
	}
	next := page[len(page)-1]
	return page, &next, nil
}

// TopologicalOrder returns an order for edges prerequisite -> dependent.
//
// Time: O(V+E). If not all nodes are emitted, there is a dependency cycle.
func TopologicalOrder(nodes []string, dependencies [][2]string) ([]string, error) {
	nodeSet := make(map[string]struct{}, len(nodes))
	for _, node := range nodes {
		nodeSet[node] = struct{}{}
	}

	adjacency := make(map[string][]string, len(nodes))
	indegree := make(map[string]int, len(nodes))
	for node := range nodeSet {
		adjacency[node] = []string{}
		indegree[node] = 0
	}
	for _, dependency := range dependencies {
		before, after := dependency[0], dependency[1]
		if _, ok := nodeSet[before]; !ok {
			return nil, errors.New("unknown prerequisite")
		}
		if _, ok := nodeSet[after]; !ok {
			return nil, errors.New("unknown dependent")
		}
		adjacency[before] = append(adjacency[before], after)
		indegree[after]++
	}

	ready := make([]string, 0)
	for node, degree := range indegree {
		if degree == 0 {
			ready = append(ready, node)
		}
	}
	sort.Strings(ready)

	result := make([]string, 0, len(nodes))
	for len(ready) > 0 {
		node := ready[0]
		ready = ready[1:]
		result = append(result, node)
		dependents := adjacency[node]
		sort.Strings(dependents)
		for _, dependent := range dependents {
			indegree[dependent]--
			if indegree[dependent] == 0 {
				ready = append(ready, dependent)
			}
		}
		sort.Strings(ready)
	}

	if len(result) != len(nodeSet) {
		return nil, errors.New("dependency graph contains a cycle")
	}
	return result, nil
}

// DataFlowNode describes governance metadata attached to a graph node.
type DataFlowNode struct {
	Name      string
	Region    string
	DataClass string
}

// FindCrossRegionViolation returns a shortest path to a disallowed region.
//
// Time: O(V+E). Space: O(V). It reports a path; policy code must decide what
// action to take.
func FindCrossRegionViolation(
	start string,
	graph map[string][]string,
	nodes map[string]DataFlowNode,
	allowedRegions map[string]bool,
) ([]string, error) {
	if _, ok := nodes[start]; !ok {
		return nil, errors.New("unknown start node")
	}

	queue := []string{start}
	parent := map[string]string{start: ""}
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		for _, neighbor := range graph[current] {
			if _, ok := nodes[neighbor]; !ok {
				return nil, errors.New("unknown graph node")
			}
			if _, seen := parent[neighbor]; seen {
				continue
			}
			parent[neighbor] = current
			if !allowedRegions[nodes[neighbor].Region] {
				path := []string{}
				for cursor := neighbor; cursor != ""; cursor = parent[cursor] {
					path = append(path, cursor)
				}
				for left, right := 0, len(path)-1; left < right; left, right = left+1, right-1 {
					path[left], path[right] = path[right], path[left]
				}
				return path, nil
			}
			queue = append(queue, neighbor)
		}
	}
	return nil, nil
}

type cacheEntry struct {
	key   string
	value string
}

// LruCache is a local, non-authoritative HashMap + doubly linked list cache.
type LruCache struct {
	capacity int
	items    map[string]*list.Element
	order    *list.List
}

// NewLruCache constructs an O(1) average get/put cache.
func NewLruCache(capacity int) (*LruCache, error) {
	if capacity <= 0 {
		return nil, errors.New("capacity must be positive")
	}
	return &LruCache{
		capacity: capacity,
		items:    make(map[string]*list.Element, capacity),
		order:    list.New(),
	}, nil
}

// Get returns a value and moves the key to the most-recent position.
func (c *LruCache) Get(key string) (string, bool) {
	element, ok := c.items[key]
	if !ok {
		return "", false
	}
	c.order.MoveToBack(element)
	return element.Value.(cacheEntry).value, true
}

// Put inserts or replaces a value, evicting the least-recent key.
func (c *LruCache) Put(key, value string) {
	if element, ok := c.items[key]; ok {
		element.Value = cacheEntry{key, value}
		c.order.MoveToBack(element)
		return
	}
	element := c.order.PushBack(cacheEntry{key, value})
	c.items[key] = element
	if c.order.Len() > c.capacity {
		oldest := c.order.Front()
		delete(c.items, oldest.Value.(cacheEntry).key)
		c.order.Remove(oldest)
	}
}

// ChannelOption is a discrete package/channel option for DP planning.
type ChannelOption struct {
	Units    int
	FeeCents int
}

// MinChannelCost computes an exact-target unbounded DP.
//
// Time: O(target * options). Space: O(target). It is a planning example, not
// a payment authorization or provider routing decision.
func MinChannelCost(target int, options []ChannelOption) (int, bool) {
	if target < 0 {
		return 0, false
	}
	infinity := int(^uint(0) >> 1)
	dp := make([]int, target+1)
	for index := 1; index <= target; index++ {
		dp[index] = infinity
		for _, option := range options {
			if option.Units <= 0 || option.FeeCents < 0 {
				return 0, false
			}
			if option.Units <= index && dp[index-option.Units] != infinity {
				candidate := dp[index-option.Units] + option.FeeCents
				if candidate < dp[index] {
					dp[index] = candidate
				}
			}
		}
	}
	if dp[target] == infinity {
		return 0, false
	}
	return dp[target], true
}
